src/main/kotlin/fr/yvbro/rexa/exception/ApiErrorReporter.kt

```diff
@@ -0,0 +1,147 @@
+            }
+            is TypeMismatchException -> {
+                return methodArgumentTypeMismatchExceptionStrategy.reportError(e as MethodArgumentTypeMismatchException)
+            }
+            is HttpMessageNotReadableException -> {
+                return messageNotReadableStrategy.reportError(e)
+            }
+            is ConstraintViolationException -> {
+                return constraintViolationExceptionStrategy.reportError(e)
+            }
+            is MissingServletRequestPartException -> {
+                return missingServletRequestPartExceptionStrategy.reportError(e)
+            }
+            else -> return e?.let { genericExceptionReportStrategy.reportError(it) }
+        }
+    }
+
+    private val genericExceptionReportStrategy: ExceptionReportStrategy<Exception> = object : ExceptionReportStrategy<Exception> {
+        override fun reportError(e: Exception): ApiErrorBean? {
+            return e.message?.let { ApiErrorBean(it, "BAD_REQUEST") }
+        }
+    }
+
+    private val bindExceptionReportStrategy: ExceptionReportStrategy<BindException> = object : ExceptionReportStrategy<BindException> {
+        override fun reportError(e: BindException): ApiErrorBean? {
+            return ApiErrorBean(INVALID_FIELD_TYPE_MESSAGE, INVALID_FIELD_TYPE, if (e.fieldError != null) e.fieldError!!.field else null)
+        }
+    }
+
+    private val methodArgumentNotValidExceptionReportStrategy: ExceptionReportStrategy<MethodArgumentNotValidException> = object : ExceptionReportStrategy<MethodArgumentNotValidException> {
+        override fun reportError(e: MethodArgumentNotValidException): ApiErrorBean? {
+            val errors = e.bindingResult.fieldErrors.stream()
+                    .map<Any> { error: FieldError -> error.defaultMessage?.let { ValidationError(error.field, it) } }
+                    .collect(Collectors.toList())
+            return ApiValidationErrorBean(VALIDATION_ERROR_MESSAGE, VALIDATION_ERROR, errors as List<ValidationError>)
+        }
+    }
+
+    private val methodArgumentTypeMismatchExceptionStrategy: ExceptionReportStrategy<MethodArgumentTypeMismatchException> = object : ExceptionReportStrategy<MethodArgumentTypeMismatchException> {
+        override fun reportError(e: MethodArgumentTypeMismatchException): ApiErrorBean? {
+            return ApiErrorBean(INVALID_FIELD_TYPE_MESSAGE, INVALID_FIELD_TYPE, e.name)
+        }
+    }
+
+    private val messageNotReadableStrategy: ExceptionReportStrategy<HttpMessageNotReadableException> = object : ExceptionReportStrategy<HttpMessageNotReadableException> {
+        override fun reportError(e: HttpMessageNotReadableException): ApiErrorBean? {
+            val causeException = e.cause
+            var message = "Required request body is missing or malformed"
+            var errorCode = "IO_ERROR"
+            var field: String? = null
+            if (causeException is JsonMappingException) {
```

Comment:
Why do we try this condition here and not after the line 91 ?  It's not better to use a else condition here ?
We set message errorCode and if is a JsonMappingException we change this value